blender/LilySurfaceScraper/Scrapers/AbstractScraper.py

```diff
@@ -166,12 +176,73 @@ def clearString(self, s):
         return ''.join(filter(lambda x: x in printable, s))
 
     def fetchVariantList(self, url):
+        # **must have self.asset_name**
+
+        # get asset name and variants
+        variants = self._fetchVariantList(url)
+        assetName = self.asset_name
+
+        root = self.getTextureDirectory(os.path.join(self.home_dir, assetName))
+
+        # download thumbnail and make metadata file if meta file is not present
+        metadataFile = os.path.join(root, ".meta")
+        if not os.path.isfile(metadataFile):
+            thumbnailName = self.downloadThumbnail(root, assetName)
+
+            metadata = {
+                "name": assetName,
+                "scraper": self.__class__.__name__,
+                "fetchUrl": url,
+                "thumbnail": thumbnailName,
+                "variants": variants,
+            }
+            with open(metadataFile, "w") as f:
+                json.dump(metadata, f, indent=4)
+        return variants
+
+    def downloadThumbnail(self, assetPath, assetName):
+        thumbnailUrl = self.getThumbnail(assetName)
+        ext = None
+        if thumbnailUrl is None:
+            print("no thumbnail found, not downloading")
+        else:
+            thumbnailReq = self._fetch(thumbnailUrl)
+            if thumbnailReq is None:
+                return None
+            thumbnailType = thumbnailReq.headers["Content-Type"]
+            if thumbnailType == 'image/png':
+                ext = "png"
+            elif thumbnailType == "image/jpeg":
+                ext = "jpg"
+            else:
+                print(f"thumbnail type '{thumbnailType}' is not a valid type")
+
+        if ext is None:
+            return None
+        thumbnailName = f"thumb.{ext}"
+        with open(os.path.join(assetPath, thumbnailName), "wb") as f:
+            f.write(thumbnailReq.content)
+        return thumbnailName
+
+    def _fetchVariantList(self, url):
         """Get a list of available variants.
         The list may be empty, and must be None in case of error."""
         raise NotImplementedError
-    
+
     def fetchVariant(self, variant_index, material_data):
         """Fill material_data with data from the selected variant.
         Must fill material_data.name and material_data.maps.
         Return a boolean status, and fill self.error to add error messages."""
         raise NotImplementedError
+
+    def getThumbnail(self, assetName):
+        """Function for getting a thumnbail for the texture, preferably using only the assetName (asset_name)
+         but you can pass more arguments with self.*** as its called after _fetchVariantList
+         returns: url
+         """
+        raise NotImplementedError
```

Comment:
Is it an error not to implement getThumbnail in a scraper? Maybe it could fail gracefully and just not display thumbnail if there is no getThumbnail available.

Reply:
this is just a function for getting the thumbnail and should be implemented by scrapers. if there is no thumbnail or you can't get one then the function should return None

https://github.com/eliemichel/LilySurfaceScraper/pull/171/commits/c5d97857ffe99f360b6ae55f0d544d50fe939961#diff-d39421f46a5e9762efbbf97b784b904efe7dc6a44c7b4570251e3f1aabaf80faL71-L72